calyx/src/passes/inliner.rs

```diff
@@ -118,35 +98,40 @@ fn inline_hole(st: &mut StructureGraph, hole: String) {
         guard_opt.map(|guard| {
             guard_map.insert(
                 st.to_atom((dest_idx, ed.dest.port_name().clone())),
-                (idx, guard),
+                guard,
             )
         });
     }
 
-    // store for all the edges that were actually inlined so that we can remove them later
-    let mut edges_inlined: Vec<EdgeIndex> = vec![];
-
     // iterate over all edges to replace holes with an expression
-    for edidx in st.edge_idx().detach() {
-        let mut ed_data = st.get_edge_mut(edidx);
-        // for the guard, recurse down the guard ast and replace any leaves with
+    for ed_idx in st.edge_idx().detach() {
+        let mut ed_data = st.get_edge_mut(ed_idx);
+        // for the guard, recur down the guard ast and replace any leaves with
         // the expression in `guard_map` adding the corresponding edges to `edges_inlined`
-        ed_data.guard = ed_data.guard.as_ref().map(|guard| {
-            tree_walk(guard.clone(), &guard_map, &mut edges_inlined)
-        });
+        ed_data.guard = ed_data
+            .guard
+            .as_ref()
+            .map(|guard| tree_walk(guard.clone(), &guard_map));
     }
 
-    // remove all the edges that have been inlined
-    for idx in edges_inlined.iter().unique() {
-        st.remove_edge(*idx);
+    // remove all the edges that have no reads from them.
+    for idx in st
+        .edge_idx()
+        .with_direction(DataDirection::Write)
+        .with_node_type(NodeType::Hole)
+        .with_port(hole)
+        .detach()
+    {
+        st.remove_edge(idx);
     }
 
     // flatten groups (maybe temporary)
     for idx in st.edge_idx().detach() {
         st.get_edge_mut(idx).group = None;
     }
     st.groups = HashMap::new();
-    st.groups.insert(None, st.edge_idx().collect());
+    st.groups
+        .insert(None, (HashMap::new(), st.edge_idx().collect()));
```

Comment:
did you add a test to make sure that this is correct? It should be but I remember I did the `edges_inlined` thing for a particular reason. Ofc I don't remember what it was exactly. Curse myself for not writing this down then...

Reply:
I can try adding one.